src/tactic/fix_by_cases.lean

```diff
@@ -0,0 +1,36 @@
+/-
+Copyright (c) 2020 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner
+-/
+
+/-!
+# Fix the `by_cases` tactic
+
+The core `by_cases` tactic has several bugs:
+ - It only works if the proposition is decidable.
+ - It sometimes unfolds the proposition.
+
+We override the `by_cases` tactic with a correct implementation here.
```

Comment:
Why not fix this in core?

Reply:
I'll move it to core once it works in mathlib.  But seeing the problems (and unforeseen issues) that I encountered while "improving" the tactic, it is absolutely necessary to test this on mathlib first.